lib/matplotlib/figure.py

```diff
@@ -1576,6 +1576,94 @@ def subplots(self, nrows=1, ncols=1, sharex=False, sharey=False,
             return axarr
 
+    def build_grid(self, layout, *, subplot_kw=None, gridspec_kw=None):
+        """Build a layout based on an grid of strings.
+
+        This is a helper function to build complex GridSpec layouts visually.
+
+        Parameters
+        ----------
+        layout : list of list of str
+
+            A visual layout of how you want your Axes to be arranged
+            labeled as strings.  For example ::
+
+               x = [['A', 'A', 'B'],
+                    ['C', 'D', 'B']]
+
+            Produces 4 axes:
+
+            - 'A' which is 1 row high and spans the first two columns
+            - 'B' which is 2 rows high and is on the right edge
+            - 'C' which in 1 row and 1 column wide in the bottom left
+            - 'D' which in 1 row and 1 column wide in the bottom center
+
+        subplot_kw : dict, optional
+            Dict with keywords passed to the
+            :meth:`~matplotlib.figure.Figure.add_subplot` call used to create
+            each subplot.
+
+        gridspec_kw : dict, optional
+            Dict with keywords passed to the
+            `~matplotlib.gridspec.GridSpec` constructor used to create
+            the grid the subplots are placed on.
+
+        Returns
+        -------
+        dict[str, Axes]
+           A dictionary mapping the string labels to the new Axes objects.
+
+        """
+        subplot_kw = subplot_kw or {}
+        gridspec_kw = gridspec_kw or {}
+
+        def _process_layout(layout):
+            unique_ids = set()
+            nested = {}
+            for j, row in enumerate(layout):
+                for k, v in enumerate(row):
+                    if isinstance(v, (str, Axes)):
+                        unique_ids.add(v)
+                    else:
+                        nested[(j, k)] = v
+
+            return unique_ids, nested
+
+        def _do_layout(gs, layout, unique_ids, nested):
+            rows, cols = layout.shape
+            covered = np.zeros((rows, cols), dtype=bool)
+            output = dict()
+            for name in unique_ids:
+                indx = np.stack(np.where(layout == name)).T
+                start_row, start_col = np.min(indx, axis=0)
+                end_row, end_col = np.max(indx, axis=0) + 1
+                slc = (slice(start_row, end_row), slice(start_col, end_col))
+                if not np.all(covered[slc] == False):
+                    raise ValueError
+                covered[slc] = True
+                if isinstance(name, str):
+                    output[name] = self.add_subplot(gs[slc], **subplot_kw)
+                else:
+                    ax = name
+                    ax.set_subplotspec(gs[slc])
+                    ax.update_params()
+                    ax.set_position(ax.figbox)
+
+            for (j, k), layout in nested.items():
+                layout = np.asarray(layout)
+                rows, cols = layout.shape
+                gs_n = GridSpecFromSubplotSpec(rows, cols, gs[j, k])
```

Comment:
could do `gs[j, k]=subplotspec(rows, cols)`.

Reply:
I don't understand this suggestion.

Reply:
Sorry.  `gs_n = gs[j, k].subgridspec(rows, cols)`

Reply:
ah, I get it now.,